### PR TITLE
Add print report functionality

### LINKED ISSUE #50: Print Advisor Engagement Report
Advisors must sometimes have a paper copy of their engagement reports for meetings or discussions.  The favored solution would be to make an engagement report into a PDF file and then print the PDF file.  Another solution would be to have a print function to simply print out the report.
Reference item 51

### PR DESCRIPTION
**Pull request by cemalettin-simsoft**
Added report printing functionality for all users. The default report page "reports/:uuid" now has a "Summary/Print View" button which shows the printable version of the report in a nice, document-like format. In that "Summary/Print View", "Print" and "Detailed View" buttons are also added. "Print" button opens the native browser print dialogue popup for printing. "Detailed View" returns to the default page. 

The rationale for not making the "Print View" a new page was to not waste another request when we had the report data available already.

Also the reason why there is a new view and not just one print button to do all the work is that the "window.print()" api to work, we need printable, nicely laid out content to be rendered on the page.

Currently, print view has all the report fields that was showing on default page **except** these fields:

- Advisor Organization
- Assessments (optional / checkbox)
- Comments
- Duration
- Principal Organization
- Sensitive Information
- Tags
- Workflow (optional / checkbox)

Still working on:

- ~~Writing tests~~
- ~~Fixing the issue when banner is large, it overlaps with other content~~

### Release notes

Closes #50 

#### User changes
- Print View and Print Functionality for reports

#### Super User changes
- Print View and Print Functionality for reports

#### Admin changes
- 

#### System admin changes
-
- [X] anet.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [X] Described the user behavior in PR body
  - [X] Referenced/updated all related issues
  - [X] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [X] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [ ] Added and/or updated unit tests
  - [X] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [X] Resolved all build errors and warnings
  - [ ] Opened debt issues for anything not resolved here

### CODE REVIEW

**Review by VassilIordanov — CHANGES_REQUESTED** [review]
ANET does currently use Minimal.js for printing. Even in this branch, if I do a CTRL+P on a report's page, it will render Minimal.js.
It will be good to not have an extra file for that. We should keep either one. Also, it may be more appropriate to not call it Print as it can be used for more than printing (pdf exports, maybe email content, etc).
Also, it would be good to have a routing entry point for that page

**Comment by cemalettin-simsoft** [comment]
> ANET does currently use Minimal.js for printing. Even in this branch, if I do a CTRL+P on a report's page, it will render Minimal.js.


I just edited Minimal.js and did a "CTRL+P" on a report page, it didn't reflect the edits i made. So Minimal.js is not used anywhere and it looks like Minimal.js is just an unreachable page (without manually writing the url with "reports/:uuid/min"). Do you mean that Minimal.js is the representation of what should be printed ? Because it is Show.js that gets printed with CTRL+P at "reports/:uuid" page.

**Comment by cemalettin-simsoft** [comment]
Also Minimal.js doesn't have custom fields which we would like to see in a report i presume. I suggest we should remove "Minimal.js" since it is being used no where, and rename "Print.js" to "Compact.js" or anything better suggested. And add as a new page "reports/:uuid/compact" that is triggered from "~~Print View~~ / Compact View" button.

**Comment by gjvoosten** [comment]
> ANET does currently use Minimal.js for printing. Even in this branch, if I do a CTRL+P on a report's page, it will render Minimal.js.

No, ANET has a `print` style in `index.css` that hides some aspects when printing _any_ ANET page.

**Comment by gjvoosten** [comment]
>So Minimal.js is not used on CTRL+P or anywhere, it only exists in Routing.js and it looks like Minimal.js is just an unreachable page (without manually writing the url with "reports/:uuid/min").

`Minimal.js` is used (linked to) in the daily rollup email (see `src/main/resources/emails/rollup.ftlh`).

**Comment by cemalettin-simsoft** [comment]
> > So Minimal.js is not used on CTRL+P or anywhere, it only exists in Routing.js and it looks like Minimal.js is just an unreachable page (without manually writing the url with "reports/:uuid/min").
> 
> `Minimal.js` is used (linked to) in the daily rollup email (see `src/main/resources/emails/rollup.ftlh`).

Ah, forgot to change my search settings from searching only ./client folder. Okay, then i can link "Compact.js" in that file also.

**Comment by gjvoosten** [comment]
> `Minimal.js` is used (linked to) in the daily rollup email (see `src/main/resources/emails/rollup.ftlh`).

Note: it's always been a pain to keep `client/src/pages/reports/{Show,Minimal}.js` somewhat in sync (it's overlooked by most), so I'd be happy to ditch `Minimal.js` (and the link to it in the rollup email) for something else/better. I'm sure `Minimal.js` had its purpose at some point in the early beginnings of ANET (before my time), but just pointing to `Show.js` would improve maintainability already.

**Comment by cemalettin-simsoft** [comment]
> > `Minimal.js` is used (linked to) in the daily rollup email (see `src/main/resources/emails/rollup.ftlh`).
> 
> Note: it's always been a pain to keep `client/src/pages/reports/{Show,Minimal}.js` somewhat in sync (it's overlooked by most), so I'd be happy to ditch `Minimal.js` (and the link to it in the rollup email) for something else/better. I'm sure `Minimal.js` had its purpose at some point in the early beginnings of ANET (before my time), but just pointing to `Show.js` would improve maintainability already.

Yes, i first looked around if i can use "Show.js" in this new page, but trying to reuse "Show.js" seemed like it would add magnitudes of more complexity than sync work. Because, for a print page, and having repetitive headers and footers on that print page, i couldn't find any other way but making it a table layout (thead and tfoot automatically turns into header and footer in print). For example, each field and fields of fields (array of objects needs to recursively form table layout) needs to compatible with table layout table>tbody>tr>th|td etc...

**Comment by gjvoosten** [comment]
> Yes, i first looked around if i can use "Show.js" in this new page, but trying to reuse "Show.js" seemed like it would add magnitudes of more complexity than sync work. […]

Sorry, I wasn't referring to the print page (that should amost by necessity be a separate thing), just to the redundancy that is `Minimal.js`. I'll open a debt issue just for that, and you can carry on with the print view as you were. :wink:

**Comment by cemalettin-simsoft** [comment]
Changed Print.js to Compact.js and added as a new page with entry point ("/:uuid/compact").

print-report-ui-31-August

Note: Only in draft version of a report, a watermark appears at the background

**Comment by VassilIordanov** [comment]
> > ANET does currently use Minimal.js for printing. Even in this branch, if I do a CTRL+P on a report's page, it will render Minimal.js.
> 
> No, ANET has a `@media print` style in `index.css` that hides some aspects when printing _any_ ANET page.

ah, that's where my confusion comes from

File: client/src/pages/reports/Show.js [hunk 1]
```diff
@@ -376,6 +374,14 @@ const ReportShow = ({ setSearchQuery, pageDispatchers }) => {
             {canEmail && (
               <Button onClick={toggleEmailModal}>Email report</Button>
             )}
+            <Button
+              value="compactView"
+              type="button"
+              bsStyle="primary"
+              onClick={onCompactClick}
+            >
+              {Settings.fields.report.compactView}
```
Comment (VassilIordanov) [inline client/src/pages/reports/Show.js:383]: It is assumed here that `fields.report.compactView` is defined, however the yml schema does not require it. Either make the field required or provide a default value for it.

File: client/src/pages/reports/Compact.js [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+            Print
+          </Button>
+        )}
+        <Button
+          value="detailedView"
+          type="button"
+          bsStyle="primary"
+          onClick={returnToDefaultPage}
+        >
+          Detailed View
```
Comment (VassilIordanov) [inline client/src/pages/reports/Compact.js:789]: Should use `Settings.fields.report.detailedView`

File: client/src/pages/reports/Show.js (hunk 1 above)
Comment (cemalettin-simsoft) [reply client/src/pages/reports/Show.js]: I wasn't awar of the required fields in that schema, changed!

File: client/src/pages/reports/Compact.js [hunk 3]
```diff
@@ -725,7 +725,7 @@ const CompactViewHeader = ({
   return (
     <header css={HEADER_STYLE}>
       <h3 css={HEADER_TITLE_STYLE} value="title">
-        {Settings.fields.report.compactView}
+        {Settings.fields.report.compactView || "Summary / Print"}
```
Comment (VassilIordanov) [inline client/src/pages/reports/Compact.js:728]: These defaults introduce repeated identical strings. I suggest to either make these fiends required in the schema or to have another mechanism dealing with the defaults

**Comment by VassilIordanov** [comment]
in general, the table layout does not adjust very well when names are long:
image
it is also quite extreme with he attendees table in IE11 - the table headers remain quite small while the rest of the table is empty

**Comment by cemalettin-simsoft** [comment]
> in general, the table layout does not make good use of space when names are long:
> image
> it is also quite extreme with he attendees table in IE11 - the table headers remain quite small while the rest of the table is empty

I thought, generally, the data column would be large and needing space. So i intentionally made headers small width. That can be changed easily.

File: client/src/pages/reports/Compact.js [hunk 4]
```diff
@@ -0,0 +1,1012 @@
...
+          <LinkTo modelType="Location" model={report.location} />
+        )}
+      </>
+    )
+  }
+
+  function getReportSubTitle() {
+    const timeToShow = report.isDraft()
+      ? moment(report.updatedAt)
+      : moment(report.releasedAt)
```
Comment (ysf-simsoft) [inline client/src/pages/reports/Compact.js:414]: When report is submitted but not published (i.e. pending approval) date is displayed as `Authored at Invalid date` because `report.releasedAt` is `null`. I suggest displaying `report.updatedAt` when report is not published.

```
    const timeToShow = !report.isPublished()
      ? moment(report.updatedAt)
      : moment(report.releasedAt)
```
Or a more informative approach would be to display `createdAt` `updatedAt` together when not published:

> Authored: 16 November 2020 @ 10:27 Last Updated: 16 November 2020 @ 12:44  [Draft]

File: client/src/pages/reports/Compact.js [hunk 5]
```diff
@@ -0,0 +1,1012 @@
...
+      <RowLabel topHeader={topHeader}>{lowerLabel}</RowLabel>
+      <RowContent>{content}</RowContent>
+    </CustomStyled>
+  )
+}
+
+CompactRow.propTypes = {
+  label: PropTypes.node,
+  content: PropTypes.node,
+  rowType: PropTypes.string
```
Comment (ysf-simsoft) [inline client/src/pages/reports/Compact.js:990]: Can be more specific such as `rowType: PropTypes.oneOf(Object.values(ROW_TYPES))`

File: client/src/pages/reports/Compact.js [hunk 6]
```diff
@@ -0,0 +1,1012 @@
...
+  font-style: italic;
+  color: grey;
+  max-width: 50%;
+  font-weight: 300;
+  width: ${props => (props.topHeader ? "15%" : "auto")};
+`
+
+const RowContent = styled.td`
+  padding: 4px 1rem;
+`
```
Comment (ysf-simsoft) [inline client/src/pages/reports/Compact.js:1010]: Field labels and field values are not horizontally aligned with row, especially for custom fields. I suggest adding;
```
  /* fixes horizontal alignment for simple custom fields such as text, number, date */
  > div.form-control-static {
    padding-top: 0;
  }
  /* fixes horizontal alignment for complex custom fields such as array-of-objects */
  div.table-responsive > table > tbody > tr:first-of-type > td {
    padding-top: 0;
  }
```
These changes doesn't perfectly align everything but it is the smallest change with maximum gain and no/minimum side-effects I could come up with.

**Review by ysf-simsoft — CHANGES_REQUESTED** [review]

File: client/src/pages/reports/Compact.js [hunk 7]
```diff
@@ -0,0 +1,1012 @@
...
+          model={Report.getPrimaryAttendee(
+            report.attendees,
+            Person.ROLE.ADVISOR
+          )}
+        />
+        <br />
+        on{" "}
+        {moment(report.engagementDate).format(
+          Report.getEngagementDateFormat()
+        )}{" "}
```
[the remark this answers is not in the document]
Comment (cemalettin-simsoft) [reply client/src/pages/reports/Compact.js:402]: Duration was intentionally left out as we want to make it as much compact as possible, it was a suggestion from @VassilIordanov. How do you feel about this change?

File: client/src/pages/reports/Compact.js (hunk 4 above)
Comment (cemalettin-simsoft) [reply client/src/pages/reports/Compact.js:414]: I think before a report is published, it may not be important to distinguish creation date vs update date. @VassilIordanov i will go with this:
```
    const timeToShow = !report.isPublished()
      ? moment(report.updatedAt)
      : moment(report.releasedAt)
```
unless you have preference

File: client/src/pages/reports/Compact.js (hunk 5 above)
Comment (cemalettin-simsoft) [reply client/src/pages/reports/Compact.js:990]: I have logic where if none of the ROW_TYPES supplied, we go with general row styling. So if i want to go with normal row styling, i pass nothing, which would fail here

File: client/src/components/ReportWorkflow.js [hunk 8]
```diff
@@ -187,3 +188,39 @@ export const ReportCompactWorkflow = ({ workflow }) => (
...
+    />
+  </Fieldset>
+)
+
+CompactRowReportWorkflow.propTypes = {
+  workflow: PropTypes.array.isRequired,
+  compactStyle: PropTypes.string
+}
+
+const CompactRowReportAction = ({ action }) => {
```
Comment (ysf-simsoft) [inline client/src/components/ReportWorkflow.js:216]: Instead of defining a new component it would be better to use existing `ReportAction` component by adding a prop `hideDetails` which defaults to `false`.

File: client/src/components/ReportWorkflow.js [hunk 9]
```diff
@@ -187,3 +188,39 @@ export const ReportCompactWorkflow = ({ workflow }) => (
...
   workflow: PropTypes.array.isRequired
 }
+
+export const CompactRowReportWorkflow = ({ workflow, compactStyle }) => (
+  <Fieldset
+    className="workflow-fieldset compact"
+    title="Workflow"
+    compactStyle={compactStyle}
+  >
+    <CompactRow
```
Comment (ysf-simsoft) [inline client/src/components/ReportWorkflow.js:198]: `Fieldset` already displays the content within a `CompactRow`. Could it be possible to re-use existing `ReportCompactWorkflow` with a little tweak? Which may also create a possibility to get rid of `ROW_TYPES.onlyData` as it is only used in here.

**Review by ysf-simsoft — CHANGES_REQUESTED** [review]

File: client/src/components/ReportWorkflow.js [hunk 10]
```diff
@@ -2,6 +2,7 @@ import Fieldset from "components/Fieldset"
 import LinkTo from "components/LinkTo"
 import _isEmpty from "lodash/isEmpty"
 import moment from "moment"
+import { CompactRow, ROW_TYPES } from "pages/reports/Compact"
```
[the remark this answers is not in the document]
Comment (cemalettin-simsoft) [reply client/src/components/ReportWorkflow.js:5]: Hmm, it might a babel-transpilation or webpack-bundle behind the scenes. I will fix this.

File: client/src/components/ReportWorkflow.js (hunk 9 above)
Comment (cemalettin-simsoft) [reply client/src/components/ReportWorkflow.js:198]: If some field is using a Fieldset, it means it has fields of fields ( I map Fieldset to table>tbody, only its children fields are rendered with CompactRow)
So conceptually:
`Fieldset(title, children_with_fields) maps to: CompactRow (th, td > table > tbody > each child tr > th = each_child_fieldname + td = each_child_content)`

File: client/src/components/ReportWorkflow.js (hunk 8 above)
Comment (cemalettin-simsoft) [reply client/src/components/ReportWorkflow.js:216]: I tried this method first, the component becomes much more complicated with conditional returns etc. I need to change 3 components and each needs to return compact version conditionally etc. I thin having an extra component here will be fine.

File: anet-dictionary.yml [hunk 11]
```diff
@@ -202,6 +202,8 @@ fields:
     nextSteps: Next steps
     keyOutcomes: Key outcomes
     reportText: Key details
+    detailedView: Detailed View
+    compactView: Summary / Print
```
Comment (gjvoosten) [inline anet-dictionary.yml:206]: These are not report fields, just buttons, so they don't belong here.
I'd say it is just fine to hard-code these into Show.js and CompactView.js.

File: src/main/resources/anet-schema.yml [hunk 12]
```diff
@@ -420,7 +420,7 @@ properties:
       report:
         type: object
         additionalProperties: false
-        required: [intent,atmosphere,atmosphereDetails,cancelled,nextSteps,reportText]
+        required: [intent,atmosphere,atmosphereDetails,cancelled,nextSteps,reportText, compactView, detailedView]
```
Comment (gjvoosten) [inline src/main/resources/anet-schema.yml:423]: Revert this change.

File: src/main/resources/anet-schema.yml [hunk 13]
```diff
@@ -454,6 +454,12 @@ properties:
             type: string
             title: The label for report's engagement details
             description: Used in the UI where report's engagement details are shown.
+          compactView:
+            type: string
+            title: The title for compact view of report for printing etc.
+          detailedView:
+            type: string
+            title: The title for detailed, default view of report
```
Comment (gjvoosten) [inline src/main/resources/anet-schema.yml:462]: Revert this change.

File: client/package.json [hunk 14]
```diff
@@ -188,8 +188,8 @@
     "test": "echo 'Run \"test-all\" to execute all client-side tests, or pick a specific test target: \"test-jest\", \"test-wdio\", \"test-wdio-ie\", \"test-e2e\".' ; exit 1",
     "test-all": "yarn run test-jest && yarn run test-wdio && yarn run test-wdio-ie && yarn run test-e2e",
     "test-e2e": "export SERVER_URL=${SERVER_URL:-'http://localhost:8180'} ; NODE_ENV=test node $(npm bin)/ava --config ava.config.js",
-    "test-wdio": "export SERVER_URL=${SERVER_URL:-'http://localhost:8180'} ; NODE_ENV=test wdio --baseUrl ${SERVER_URL} ./config/wdio.config.js",
     "test-jest": "export ANET_DICTIONARY=../anet-dictionary.yml ; NODE_ENV=test jest --testPathPattern='tests/(actions|reducers|models|utils)'",
+    "test-wdio": "export SERVER_URL=${SERVER_URL:-'http://localhost:8180'} ; export ANET_DICTIONARY=../anet-dictionary.yml; NODE_ENV=test wdio --baseUrl ${SERVER_URL} ./config/wdio.config.js",
```
Comment (gjvoosten) [inline client/package.json:192]: wdio tests should get the dictionary through a request to the server, not through a local file; revert this change.

File: client/src/index.css [hunk 15]
```diff
@@ -1097,4 +1101,8 @@ header.searchPagination {
 		display: block !important;
 		height: 100% !important;
 	}
+	/* adds random characters to printed pdf*/
```
Comment (gjvoosten) [inline client/src/index.css:1104]: Should at least have a space before the comment close ` */`, but frankly I don't understand this comment…

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/pages/reports/Compact.js [hunk 16]
```diff
@@ -0,0 +1,939 @@
...
+  const { currentUser, appSettings } = useContext(AppContext)
+  return (
+    <FooterContentS bgc={appSettings[SETTING_KEY_COLOR]}>
+      <img src={anetLogo} alt="logo" width="50" height="12" />
+      <ClassificationBanner />
+      <PrintedByBoxS>
+        <div>
+          printed by <LinkTo modelType="Person" model={currentUser} />
+        </div>
+        <div>{moment.utc().toString()}</div>
```
Comment (gjvoosten) [inline client/src/pages/reports/Compact.js:788]: Please format the timestamp according to the format from the settings (e.g. `Settings.dateFormats.forms.displayLong.withTime`).

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/index.css (hunk 15 above)
Comment (cemalettin-simsoft) [reply client/src/index.css:1104]: react-svg library was adding some elements to the end of the document which was causing empty space => empty pages.

File: anet-dictionary.yml (hunk 11 above)
Comment (cemalettin-simsoft) [reply anet-dictionary.yml:206]: I also think it is fine, there were user feedback to change the previous hard-coded labels, and i thought might as well give admins a way to change when needed. If @VassilIordanov doesn't have a strong requirement, i can revert or put it somewhere else.

File: client/src/pages/reports/Compact.js [hunk 17]
```diff
@@ -0,0 +1,939 @@
...
+              />
+            )
+          })}
+        </tbody>
+      </table>
+    )
+  }
+  // first item is primary, grouped around that item
+  // people without organization at the end
+  function sortGroupByOrganization(attendees) {
```
Comment (ysf-simsoft) [inline client/src/pages/reports/Compact.js:584]: People without organization doesn't actually sorted to the end. In the following screenshot `OF-2 DVISOR, A` and `CIV HUNTMAN, Hunter` doesn't have a position.

image

I also think that sorting logic here can be simplified. Following patch file contains my suggestion.

[attendess-sort.patch.txt](https://github.com/NCI-Agency/anet/files/5589069/attendess-sort.patch.txt)

File: client/src/pages/reports/Compact.js [hunk 18]
```diff
@@ -0,0 +1,939 @@
...
+            prevDiffOrgName = renderOrgName
+              ? attendee.position?.organization?.shortName
+              : prevDiffOrgName
+            return (
+              <CompactRow
+                key={attendee.uuid}
+                label={
+                  <>
+                    <LinkTo modelType="Person" model={attendee} />
+                    {renderOrgName && (
```
Comment (ysf-simsoft) [inline client/src/pages/reports/Compact.js:539]: It would be better to display ` 'NA'` for all people without an organization. Such that, `{(renderOrgName || !attendee.position?.organization) && (`

**Review by ysf-simsoft — CHANGES_REQUESTED** [review]

**Review by ysf-simsoft — APPROVED** [review]
